### DISCUSSION TITLE
"Next day starts at" Option

### DISCUSSION THREAD

**Discussion by bradenpecora** [fabd/kanji-koohii#248] [1. Feedback & Feature Requests]
Hi Fabrice,

As always, huge fan of the site. I appreciate your work very much.

--

As someone who frequently goes to bed past midnight, I often have to remember to do my SRS kanji reviews before midnight. If I forget to do that, I'm forced to combine two days of SRS.

Anki has an option called "Next day starts at ___ hours past midnight" that allows you to choose when the next SRS day starts at. This would be a very useful feature to me that doesn't seem to difficult to implement.

If you like this feature but don't have the time to implement it, I could try my hand at creating a pull request... I'm a bit of a novice, but I'll try me best!

Thanks!

**Comment by fabd** [comment]
Hi braden

The functionality is there, but not in an obvious way : you'll want to tweak the **Timezone** setting in Account > Edit Account.

> Due flashcards appear each day at midnight local time. Adjust the timezone to your local time, or move it forward/backward if you'd like for due cards to appear at another time of the day.

Timezone setting isn't actually used for anything else but setting flashcard review timestamps.

To get `Next day starts at +X hours past midnight` you'd want to shift your actual timezone back X hours, in Edit Account.

This should cause [the conditional](https://github.com/fabd/kanji-koohii/blob/master/src/lib/peer/ReviewsPeer.php#L577) in the due card selection to think you're late evening when you're actually past midnight (if you shift your timezone -X hours).

Honestly handling of dates melts my brain and unfortunately this is the first and only real php site I've ever built, started way back in 2005 or so... so there are lots of problems with the way the site handle timestamps.

The only reason this works right now is in a very confusing way, it compares the user's localtime (which is derived from server time + server time zone setting + user's timezone setting) to the `expiredate` which is stored as a DATE. expiredate was also derived from user's local time, but since it truncates HMS, it is the user's midnight time.

You can try to add an actual setting if you like, but it's not trivial as a lot of the code is really outdated and doesn;'t follow modern php framework practices.

It would look something like this:

- add a `Next day starts at ___ hours past midnight` setting in the Spaced Repetition Screen, which is a Vue component, https://github.com/fabd/kanji-koohii/blob/master/src/vite/src/vue/account/SpacedRepetitionForm.vue
- the form is instanced from this php template https://github.com/fabd/kanji-koohii/blob/master/src/apps/koohii/modules/account/templates/spacedrepetitionSuccess.php
- the form POST would have to be handled in https://github.com/fabd/kanji-koohii/blob/master/src/apps/koohii/modules/account/actions/actions.class.php#L472
- the option would need to be added to the UsersSettingsPeer db model https://github.com/fabd/kanji-koohii/blob/master/src/lib/peer/UsersSettingsPeer.php
- the mysql schema updated with an incremental sql file in this folder https://github.com/fabd/kanji-koohii/blob/master/src/data/schemas/incremental/README.md
- I would probably add a `sqlMidnightTime` to UsersPeer, as I';m not sure all the other places `sqlLocalTime` is used and what side effects it would have to change the existing method https://github.com/fabd/kanji-koohii/blob/master/src/lib/peer/UsersPeer.php#L309 - this would essentially do sql `DATE( x )` of what the existing function returnsm truncating the HMS, so it is a localized midnight time
- then use that in ReviewsPeer::getFLashcardsForReview https://github.com/fabd/kanji-koohii/blob/master/src/lib/peer/ReviewsPeer.php#L530  instead of sqlLocalTime, - therefore then the midnight time used in the due card selection would take in account the user's `Next day starts at` setting

> If you like this feature but don't have the time to implement it, I could try my hand at creating a pull request

I'll be busy for a while with the dashboard redesign, so it might take a while till I get to it myself.

So If you'd like to take a stab at this, you're welcome to do a PR.

I didnt mean to be handholding, just that you;'re looking at a lot of questionable code due to maintaining a very old codebase, AND also you can't do thing even the "symfony" way as eg. it's using a custom Zend-Db like API instead of Sf's "Doctrine".

So I've broke down basically the simplest way I can see how to plug this functionality without refactoring too much stuff.

Either way thanks for making me think through  this again. 

This would close   https://github.com/fabd/kanji-koohii/issues/57

(converting to an issue, as it is actionable)

**Reply by bradenpecora** [reply]
Wow, great response. Hand holding is what I would need to tackle something like this. I've got sometime during my spring break, so I'll take a look at it then. I appreciate it!

**Reply by fabd** [reply]
Ok great thanks braden! I will focus on the dashboard then. Keep me updated and don't hesitate to ask if there's issues with the repo. Afaik, the setup instructions are up to date. You can reach me on the site's contact form. Cheers

**Reply by joseph-allen** [reply]
Thanks!
